<a href="https://colab.research.google.com/github/jchoi92k/CS8980/blob/main/Homework_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Joon Suh Choi 002-51-1521

1. Using NLTK tokenize all documents, separated by polarity, remove stop words, and list
the top 20 most frequent tokens (and their counts) for the positive reviews, and the top
20 most frequent tokens (and their counts). What kind of things do you notice are
different between the two sets? (30 points)


In [ ]:
!tar -xvf /content/review_polarity.tar.gz

In [2]:
from sklearn.datasets import load_files
import numpy as np

# Load files and labels
data = load_files('txt_sentoken', encoding='utf-8')
data_neg = load_files('txt_sentoken', categories = 'neg', encoding='utf-8')
data_pos = load_files('txt_sentoken', categories= 'pos', encoding='utf-8')
data_pos.target = np.asarray([1 for w in range(len(data_pos.target))])

In [3]:
import nltk
from nltk.corpus import stopwords  
from nltk.tokenize import word_tokenize  

nltk.download('stopwords')
nltk.download('punkt')

stop_words = set(stopwords.words('english'))  

tokens_neg = []
tokens_pos = []

for text in data_neg.data:
  word_tokens = word_tokenize(text)  
  filtered_sentence = [w for w in word_tokens if not w in stop_words]
  tokens_neg += filtered_sentence

for text in data_pos.data:
  word_tokens = word_tokenize(text)  
  filtered_sentence = [w for w in word_tokens if not w in stop_words]
  tokens_pos += filtered_sentence

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [4]:
from collections import Counter

# For positive tokens
Counter(tokens_pos).most_common(20)

[(',', 42448),
 ('.', 33714),
 ("'s", 9473),
 ('``', 8494),
 (')', 6039),
 ('(', 6014),
 ('film', 5186),
 ('one', 2943),
 ("n't", 2775),
 ('movie', 2497),
 ('like', 1713),
 ('?', 1570),
 (':', 1502),
 ('story', 1231),
 ('also', 1200),
 ('good', 1190),
 ('even', 1175),
 ('time', 1171),
 ('would', 1079),
 ('character', 1067)]

In [5]:
# For negative tokens
Counter(tokens_neg).most_common(20)

[(',', 35269),
 ('.', 32162),
 ('``', 9123),
 ("'s", 8655),
 (')', 5742),
 ('(', 5650),
 ('film', 4257),
 ("n't", 3442),
 ('movie', 3174),
 ('one', 2637),
 ('?', 2201),
 ('like', 1832),
 (':', 1540),
 ('even', 1381),
 ('would', 1185),
 ('good', 1126),
 ('time', 1111),
 ('!', 1056),
 ('get', 1039),
 ('bad', 1019)]

To more clearly see the differences between the two lists I used normalized (per 1000 words) lists which are shown below. The token [good] appeared in both lists: the raw frequency was slightly higher in pos which seemed to be as expected, but the normalized frequency revealed that [good] was actually less frequent in positive reviews (although the difference seems negligible). Exclamation point only appeared frequently in negative reviews, which probably indicates that stronger opinions were expressed in them compared to positive reviews. The word [bad] also only appeared frequently in negative reviews, which is as expected. The word [character] appeared in the positive list only, suggesting that if a review discusses characters, it is likely to be on a positive note. The token [n't] appeared more frequently in negative texts, which is somewhat expected considering that it is a contraction of the negation word [not]. The word [also] appeared in the positive list only, which suggests that either 1) something other than also was used frequently in negative reviews as transition words, or 2) maybe there was less trasition in negative reviews and criticism focused more on a single element while positive comments were spread across different elements. A more detailed analysis would prove or disprove this hypothesis.

In [6]:
# Normalized per 1000 tokens: positive reviews
for w in Counter(tokens_pos).most_common(20):
  print(f'("{w[0]}", {w[1]/len(tokens_pos) * 1000})')

(",", 86.79365650584377)
(".", 68.93519919520394)
("'s", 19.369494630603516)
("``", 17.367728005103583)
(")", 12.34797615055575)
("(", 12.296858514562391)
("film", 10.603842410462349)
("one", 6.017568109138198)
("n't", 5.674057595262827)
("movie", 5.105629483016677)
("like", 3.502580418264945)
("?", 3.2101875403829325)
(":", 3.0711475704809965)
("story", 2.517032396312987)
("also", 2.4536465276812223)
("good", 2.433199473283879)
("even", 2.4025288916878633)
("time", 2.394350069928926)
("would", 2.206237169473366)
("character", 2.1817007041965533)


In [7]:
# Normalized per 1000 tokens: negative reviews
for w in Counter(tokens_neg).most_common(20):
  print(f'("{w[0]}", {w[1]/len(tokens_neg) * 1000})')

(",", 80.13132185213796)
(".", 73.07220429863226)
("``", 20.72749579679193)
("'s", 19.66419775525969)
(")", 13.045849048030172)
("(", 12.836824646703322)
("film", 9.671922570091334)
("n't", 7.820239014858909)
("movie", 7.211341845776343)
("one", 5.991275503248966)
("?", 5.000681601308675)
("like", 4.162311991639023)
(":", 3.498886717862498)
("even", 3.1376380242650064)
("would", 2.692325169264325)
("good", 2.558276911891671)
("time", 2.5241968464579454)
("!", 2.3992366065342847)
("get", 2.3606125323760625)
("bad", 2.3151724451310947)


2. Using the code from previous lectures, build 3 polarity classifiers using the following
parameters (20 points). Note: just train the models.
a) For training: use 50% of the positive dataset and 70% of the negative dataset. For
your model use: NaiveBayes with the TF-IDF vectorizer.
b) For training: use 50% of the negative dataset and 70% of the positive dataset. For
your model use: NaiveBayes with the TF-IDF vectorizer.
c) For training: use 25% of the negative dataset and 25% of the positive dataset. For
your model use: SVM with the TF-IDF vectorizer.


In [8]:
%matplotlib inline
import os
import numpy as np
import matplotlib.pyplot as plt
import random
import sklearn.metrics
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC

random.seed(12345)

### Train Model A

# Create train and test set using train_test_split and concatenate
apx_train, apx_test, apy_train, apy_test = train_test_split(data_pos.data, data_pos.target, train_size=0.5)
anx_train, anx_test, any_train, any_test = train_test_split(data_neg.data, data_neg.target, train_size=0.7)

ax_train = apx_train + anx_train
ax_test = apx_test + anx_test
ay_train = np.concatenate((apy_train, any_train))
ay_test = np.concatenate((apy_test, any_test))

# Actual training
model_a = make_pipeline(TfidfVectorizer(), MultinomialNB())
model_a.fit(ax_train, ay_train)

Pipeline(memory=None,
         steps=[('tfidfvectorizer',
                 TfidfVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.float64'>,
                                 encoding='utf-8', input='content',
                                 lowercase=True, max_df=1.0, max_features=None,
                                 min_df=1, ngram_range=(1, 1), norm='l2',
                                 preprocessor=None, smooth_idf=True,
                                 stop_words=None, strip_accents=None,
                                 sublinear_tf=False,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, use_idf=True,
                                 vocabulary=None)),
                ('multinomialnb',
                 MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))],
         verbose=False)

In [9]:
### Train Model B

random.seed(12345)

# Create train and test set
bpx_train, bpx_test, bpy_train, bpy_test = train_test_split(data_pos.data, data_pos.target, train_size=0.7)
bnx_train, bnx_test, bny_train, bny_test = train_test_split(data_neg.data, data_neg.target, train_size=0.5)

bx_train = bpx_train + bnx_train
bx_test = bpx_test + bnx_test
by_train = np.concatenate((bpy_train, bny_train))
by_test = np.concatenate((bpy_test, bny_test))

# Actual training
model_b = make_pipeline(TfidfVectorizer(), MultinomialNB())
model_b.fit(bx_train, by_train)

Pipeline(memory=None,
         steps=[('tfidfvectorizer',
                 TfidfVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.float64'>,
                                 encoding='utf-8', input='content',
                                 lowercase=True, max_df=1.0, max_features=None,
                                 min_df=1, ngram_range=(1, 1), norm='l2',
                                 preprocessor=None, smooth_idf=True,
                                 stop_words=None, strip_accents=None,
                                 sublinear_tf=False,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, use_idf=True,
                                 vocabulary=None)),
                ('multinomialnb',
                 MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))],
         verbose=False)

In [10]:
### Train Model C

random.seed(12345)

# Create train and test set
cpx_train, cpx_test, cpy_train, cpy_test = train_test_split(data_pos.data, data_pos.target, train_size=0.25)
cnx_train, cnx_test, cny_train, cny_test = train_test_split(data_neg.data, data_neg.target, train_size=0.25)

cx_train = cpx_train + cnx_train
cx_test = cpx_test + cnx_test
cy_train = np.concatenate((cpy_train, cny_train))
cy_test = np.concatenate((cpy_test, cny_test))

# Actual training
model_c = make_pipeline(TfidfVectorizer(), SVC())
model_c.fit(cx_train, cy_train)

Pipeline(memory=None,
         steps=[('tfidfvectorizer',
                 TfidfVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.float64'>,
                                 encoding='utf-8', input='content',
                                 lowercase=True, max_df=1.0, max_features=None,
                                 min_df=1, ngram_range=(1, 1), norm='l2',
                                 preprocessor=None, smooth_idf=True,
                                 stop_words=None, strip_accents=None,
                                 sublinear_tf=False,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, use_idf=True,
                                 vocabulary=None)),
                ('svc',
                 SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None,
                     coef0=0.0, decision_function_shape=

3. Using the models from question 2, evaluate them on their individual rest of the dataset.
This is, for a) 50% positive and 30% negative, for b) 50% negative and 30% positive, and
for c) 75% negative and 75% positive. Calculate and show ONLY the following metrics for
each model: Accuracy, Precision, Recall, Macro F1-score. (15 points).

In [11]:
# Evaluate Model A
labels_a = model_a.predict(ax_test)

print('Accuracy:', sklearn.metrics.accuracy_score(ay_test, labels_a))
print('Precision:', sklearn.metrics.precision_score(ay_test, labels_a))
print('Recall:', sklearn.metrics.recall_score(ay_test, labels_a))
print('F1 Score:', sklearn.metrics.f1_score(labels_a, ay_test, average='macro'))

Accuracy: 0.3775
Precision: 1.0
Recall: 0.004
F1 Score: 0.2772081074608669


In [12]:
# Evaluate Model B
labels_b = model_b.predict(bx_test)

print('Accuracy:', sklearn.metrics.accuracy_score(by_test, labels_b))
print('Precision:', sklearn.metrics.precision_score(by_test, labels_b))
print('Recall:', sklearn.metrics.recall_score(by_test, labels_b))
print('F1 Score:', sklearn.metrics.f1_score(labels_b, by_test, average='macro'))

Accuracy: 0.375
Precision: 0.375
Recall: 1.0
F1 Score: 0.2727272727272727


In [13]:
# Evaluate Model C
labels_c = model_c.predict(cx_test)

print('Accuracy:', sklearn.metrics.accuracy_score(cy_test, labels_c))
print('Precision:', sklearn.metrics.precision_score(cy_test, labels_c))
print('Recall:', sklearn.metrics.recall_score(cy_test, labels_c))
print('F1 Score:', sklearn.metrics.f1_score(labels_c, cy_test, average='macro'))

Accuracy: 0.7833333333333333
Precision: 0.7906976744186046
Recall: 0.7706666666666667
F1 Score: 0.783298564791951


4) Using the model performance metrics from question 3, answer the following
questions. Please provide logical and intuitive rationale for your answers, simple
answers like: because it has the best score, will not be sufficient. (40 points):
a) What is the best performing model?
b) Why do you think this is the best performing model?
c) How does class imbalance play in determining polarity?
d) Do you think either more data or a better model is a better approach for this
kind of task?

a) The best performing model is Model C because it performs reasonably well in all metrics (i.e., accuracy, precision, recall, and F1 score) while the other two models excel in only one metric (i.e., precision for Model A and recall for model B).

b) Model C is the best performing model because it has a better data balance than the other models (i.e., same ratio between positive and negative data). Also, SVM may have outperformed Naive Bayes (which assumes independence between data), because the reviews are not entirely independent from one another (because a reviewer may have read another review - not necessarily for the same movie - and that may have affected the newly created review).

c) More negative data translates to higher precision and lower recall, and more positive data translates to higher recall and lower precision. The way the model is set up currently, higher precision means that more texts were labeled as negative than the true labels, and higher recall means that more texts were labeled as positive compared to the the true labels. This is expected because the model will train itself to label texts more frequently into the bigger portion of the train data set.

d) I believe building a better model could be more beneficial over collecting more data because a 1 million token corpus (for both positive and negative texts in the txt_sentoken folder) seems reasonably big enough for the task of binary classification. It could also be beneficial to look at the thresholds as well as some ambiguous data being fed to the model. Looking at the Readme file for this dataset, the labels were created using a threshold of 4/10 for negative reviews and 7/10 for positive reviews. Tinkering with these thresholds or working with the unlabeled texts to create an improved set of labels could be a better objective than procuring more data.

5) Using NLTK and VADER, calculate the sentiment score for all documents in the
positive polarity. Calculate the polarity threshold needed (and reasonable) to have the
majority of the document labels match. Do the same for the negative class. Provide the
threshold needed, the reason why you think this threshold is reasonable, and the
accuracy percentage (how many documents are correctly labeled using this threshold).
(45 points):


In [14]:
import nltk

# Download Vader
nltk.download('vader_lexicon')

[nltk_data] Downloading package vader_lexicon to /root/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


True

In [15]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer as SIA

sia = SIA()

# Get polarity scores for both positive and negative items
positems = [sia.polarity_scores(w)['compound'] for w in data_pos.data]
negitems = [sia.polarity_scores(w)['compound'] for w in data_neg.data]

/usr/local/lib/python3.7/dist-packages/nltk/twitter/__init__.py:20: UserWarning: The twython library has not been installed. Some functionality from the twitter package will not be available.
  warnings.warn("The twython library has not been installed. "


In [16]:
from sklearn.model_selection import KFold

random.seed(12345)

# If we are considering accuracy only (i.e., not precision, recall, or F1)

# For Kfold cross validation
kf = KFold(n_splits = 10)

positems=np.array(positems)
negitems=np.array(negitems)

batchn = 0
# To average the optimal thresholds found in each fold
allthresh = []

# Split data using kf.split
for (posfolds_train, posfolds_test), (negfolds_train, negfolds_test) in zip(kf.split(positems), kf.split(negitems)):
  batchn += 1
  X_train = np.concatenate((positems[posfolds_train], negitems[negfolds_train]))
  y_train =  [1 for w in range(len(posfolds_train))] + [0 for w in range(len(negfolds_train))]
  X_test = np.concatenate((positems[posfolds_test], negitems[negfolds_test]))
  y_test =  [1 for w in range(len(posfolds_test))] + [0 for w in range(len(negfolds_test))]

  # For finding the optimal threshold
  opt = 0
  thrsh = 0

  for p in np.arange(-1, 1, 0.001):
    crr = 0
    fll = len(X_train)
    for a, b in zip(X_train, y_train):
      if (a > p and b == 1) or (a < p and b == 0):
        crr += 1
      else:
        pass
    if crr/fll > opt:
      opt = crr/fll
      thrsh = p

  # Evaluating the result on the test set of each fold
  crr = 0
  fll = len(X_test)
  for a, b in zip(X_test, y_test):
    if (a > p and b == 1) or (a < p and b == 0):
      crr += 1
    else:
      pass
  opt = crr/fll
  
  allthresh.append(thrsh)
  print(f'For batch number {batchn}, threshold is: {thrsh}, and accuracy of test set lables is: {opt}')

# For final evaluation on complete set
completeset = np.concatenate((positems[posfolds_train], negitems[negfolds_train]))
completelabels = [1 for w in range(len(positems))] + [0 for w in range(len(negitems))]
finthresh = sum(allthresh)/len(allthresh)

crr = 0
fll = len(completeset)
for a, b in zip(completeset, completelabels):
  if (a > finthresh and b == 1) or (a < finthresh and b == 0):
    crr += 1
  else:
    pass
if crr/fll > opt:
  opt = crr/fll

print(f'Final threshold is {finthresh} and accuarcy of the entire dataset is {opt}')

For batch number 1, threshold is: 0.9740000000000018, and accuracy of test set lables is: 0.545
For batch number 2, threshold is: 0.9740000000000018, and accuracy of test set lables is: 0.52
For batch number 3, threshold is: 0.9740000000000018, and accuracy of test set lables is: 0.52
For batch number 4, threshold is: 0.9710000000000019, and accuracy of test set lables is: 0.555
For batch number 5, threshold is: 0.9740000000000018, and accuracy of test set lables is: 0.54
For batch number 6, threshold is: 0.9740000000000018, and accuracy of test set lables is: 0.545
For batch number 7, threshold is: 0.9710000000000019, and accuracy of test set lables is: 0.56
For batch number 8, threshold is: 0.9740000000000018, and accuracy of test set lables is: 0.525
For batch number 9, threshold is: 0.9740000000000018, and accuracy of test set lables is: 0.54
For batch number 10, threshold is: 0.9740000000000018, and accuracy of test set lables is: 0.53
Final threshold is 0.9734000000000018 and acc

The threshold I found is 0.9734. I think this is reasonable if we are only assessing accuracy because it is a threshold that has been derived using 10-fold cross validation. The accuracy percentage is 66.833. Using other measures such as g-mean or F1 score as the yardstick will result in different thresholds.

Bonus (40 points): Repeat questions 2,3 and 4 removing all stopwords. Answer the
following questions: Did this change the results in any way? Why do you think so?

In [17]:
random.seed(12345)

tokens_neg = []
tokens_pos = []

for text in data_neg.data:
  word_tokens = word_tokenize(text)  
  filtered_sentence = [w for w in word_tokens if not w in stop_words]
  tokens_neg.append(filtered_sentence)

for text in data_pos.data:
  word_tokens = word_tokenize(text)  
  filtered_sentence = [w for w in word_tokens if not w in stop_words]
  tokens_pos.append(filtered_sentence)

# post-stopword-removal texts
postext_ns = [' '.join(w) for w in tokens_pos]
negtext_ns = [' '.join(w) for w in tokens_neg]

random.seed(12345)

### Train Model A

# Create train and test set using train_test_split and concatenate
apx_train, apx_test, apy_train, apy_test = train_test_split(postext_ns, data_pos.target, train_size=0.5)
anx_train, anx_test, any_train, any_test = train_test_split(negtext_ns, data_neg.target, train_size=0.7)

ax_train = apx_train + anx_train
ax_test = apx_test + anx_test
ay_train = np.concatenate((apy_train, any_train))
ay_test = np.concatenate((apy_test, any_test))

# Actual training
model_a = make_pipeline(TfidfVectorizer(), MultinomialNB())
model_a.fit(ax_train, ay_train)

Pipeline(memory=None,
         steps=[('tfidfvectorizer',
                 TfidfVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.float64'>,
                                 encoding='utf-8', input='content',
                                 lowercase=True, max_df=1.0, max_features=None,
                                 min_df=1, ngram_range=(1, 1), norm='l2',
                                 preprocessor=None, smooth_idf=True,
                                 stop_words=None, strip_accents=None,
                                 sublinear_tf=False,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, use_idf=True,
                                 vocabulary=None)),
                ('multinomialnb',
                 MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))],
         verbose=False)

In [18]:
### Train Model B

random.seed(12345)

# Create train and test set
bpx_train, bpx_test, bpy_train, bpy_test = train_test_split(postext_ns, data_pos.target, train_size=0.7)
bnx_train, bnx_test, bny_train, bny_test = train_test_split(negtext_ns, data_neg.target, train_size=0.5)

bx_train = bpx_train + bnx_train
bx_test = bpx_test + bnx_test
by_train = np.concatenate((bpy_train, bny_train))
by_test = np.concatenate((bpy_test, bny_test))

# Actual training
model_b = make_pipeline(TfidfVectorizer(), MultinomialNB())
model_b.fit(bx_train, by_train)

Pipeline(memory=None,
         steps=[('tfidfvectorizer',
                 TfidfVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.float64'>,
                                 encoding='utf-8', input='content',
                                 lowercase=True, max_df=1.0, max_features=None,
                                 min_df=1, ngram_range=(1, 1), norm='l2',
                                 preprocessor=None, smooth_idf=True,
                                 stop_words=None, strip_accents=None,
                                 sublinear_tf=False,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, use_idf=True,
                                 vocabulary=None)),
                ('multinomialnb',
                 MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))],
         verbose=False)

In [19]:
### Train Model C

random.seed(12345)

# Create train and test set
cpx_train, cpx_test, cpy_train, cpy_test = train_test_split(postext_ns, data_pos.target, train_size=0.25)
cnx_train, cnx_test, cny_train, cny_test = train_test_split(negtext_ns, data_neg.target, train_size=0.25)

cx_train = cpx_train + cnx_train
cx_test = cpx_test + cnx_test
cy_train = np.concatenate((cpy_train, cny_train))
cy_test = np.concatenate((cpy_test, cny_test))

# Actual training
model_c = make_pipeline(TfidfVectorizer(), SVC())
model_c.fit(cx_train, cy_train)

Pipeline(memory=None,
         steps=[('tfidfvectorizer',
                 TfidfVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.float64'>,
                                 encoding='utf-8', input='content',
                                 lowercase=True, max_df=1.0, max_features=None,
                                 min_df=1, ngram_range=(1, 1), norm='l2',
                                 preprocessor=None, smooth_idf=True,
                                 stop_words=None, strip_accents=None,
                                 sublinear_tf=False,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, use_idf=True,
                                 vocabulary=None)),
                ('svc',
                 SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None,
                     coef0=0.0, decision_function_shape=

In [20]:
# Evaluate Model A
labels_a = model_a.predict(ax_test)

print('Accuracy:', sklearn.metrics.accuracy_score(ay_test, labels_a))
print('Precision:', sklearn.metrics.precision_score(ay_test, labels_a))
print('Recall:', sklearn.metrics.recall_score(ay_test, labels_a))
print('F1 Score:', sklearn.metrics.f1_score(labels_a, ay_test, average='macro'))

Accuracy: 0.3925
Precision: 1.0
Recall: 0.028
F1 Score: 0.3034804480082551


In [21]:
# Evaluate Model B
labels_b = model_b.predict(bx_test)

print('Accuracy:', sklearn.metrics.accuracy_score(by_test, labels_b))
print('Precision:', sklearn.metrics.precision_score(by_test, labels_b))
print('Recall:', sklearn.metrics.recall_score(by_test, labels_b))
print('F1 Score:', sklearn.metrics.f1_score(labels_b, by_test, average='macro'))

Accuracy: 0.3775
Precision: 0.37593984962406013
Recall: 1.0
F1 Score: 0.2772081074608669


In [22]:
# Evaluate Model C
labels_c = model_c.predict(cx_test)

print('Accuracy:', sklearn.metrics.accuracy_score(cy_test, labels_c))
print('Precision:', sklearn.metrics.precision_score(cy_test, labels_c))
print('Recall:', sklearn.metrics.recall_score(cy_test, labels_c))
print('F1 Score:', sklearn.metrics.f1_score(labels_c, cy_test, average='macro'))

Accuracy: 0.7773333333333333
Precision: 0.7549019607843137
Recall: 0.8213333333333334
F1 Score: 0.7769014144717507


a) The best performing model is still Model C for the same reasons as stated above (i.e., better overall scores including F1 Score).

b) I think Model C remains the best performing model because the two relevant variables (i.e., ratio between pos/neg data being fed to the model, and SVM over Naive Bayes) did not change.

c) Same as before: more negative data translates to higher precision and more positive data translates to higher recall.

d) I believe building a better model could be more beneficial over collecting more data because a 1 million token corpus (for both positive and negative texts in the txt_sentoken folder) for the same reasons as stated above.

While Model C is still the best-performing model out of the three, its performance metrics (with the exception of recall scores) drops slightly compared to the older Model C which was trained on a corpus pre-stopword-removal. This could either be 1) due to chance considering the size of the differences, and a different sampling with a different seed will produce different results, or 2) an indication that stop words, while considered neutral in textblob and SIA as an individual token, perhaps contains some information about the polarity of the text they appear in. For instance, it was shown from the normalized word frequency lists in Q1 that the token [would] appeared more frequently in negative reviews than in positive reviews in this particular corpus. Perhaps this has to do with writing style, or whether some tokens appear more frequently as a component of ngrams, p-frames or other phraseological units.